### PR TITLE
add roundtrip serialization tests

### PR DESCRIPTION
**Pull request by valentinsulzer**
# Description

Tests that parameter values are correct after full roundtrip. Threw lots of edge cases at it, no bugs found so far.

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/main/CHANGELOG.md) to document the change (include PR #)

# Important checks:

Please confirm the following before marking the PR as ready for review:
- No style issues: `nox -s pre-commit`
- All tests pass: `nox -s tests`
- The documentation builds: `nox -s doctests`
- Code is commented for hard-to-understand areas
- Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_parameters/test_parameter_values_serialisation.py [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+
+def _assert_evaluate_equal(pv_original, pv_loaded, name, inputs):
+    """Assert that a FunctionParameter evaluates identically
+    before and after roundtrip."""
+    fp = pybamm.FunctionParameter(name, inputs)
+    orig = pv_original.process_symbol(fp).evaluate()
+    loaded = pv_loaded.process_symbol(fp).evaluate()
+    assert not np.any(np.isnan(orig)), f"'{name}' original has NaN"
+    np.testing.assert_allclose(orig, loaded)
+
```
Comment (MarcBerliner) [inline tests/unit/test_parameters/test_parameter_values_serialisation.py]: I think we should check for exact floating point equivalence here and elsewhere for safety

**Review by MarcBerliner — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: tests/unit/test_parameters/test_parameter_values_serialisation.py (hunk 1 above)
Comment (valentinsulzer) [reply tests/unit/test_parameters/test_parameter_values_serialisation.py]: updated, except for cases with functions / solvers where we get some machine error

Comment (MarcBerliner) [reply tests/unit/test_parameters/test_parameter_values_serialisation.py]: All the tests that fail with `_assert_evaluate_equal` are good first candidates for looking into these serialization issues. I think we shouldn't see any machine error even if they go through a solver

**Review by MarcBerliner — APPROVED** [review]
